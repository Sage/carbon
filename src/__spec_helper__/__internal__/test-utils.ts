/* eslint-disable jest/no-conditional-expect */
/* eslint-disable jest/no-identical-title */
/* eslint-disable jest/no-export */
import { render } from "@testing-library/react";
import { sprintf } from "sprintf-js";
import {
  LayoutProps,
  FlexboxProps,
  BackgroundProps,
  PositionProps,
  GridProps,
} from "styled-system";

import { space } from "../../style/themes/base/base-theme.config";
import { mockMatchMedia } from "../mock-match-media";

const isUpper = (char: string) => char.toUpperCase() === char;
const humpToDash = (acc: string, char: string) =>
  `${acc}${isUpper(char) ? `-${char.toLowerCase()}` : char}`;

const toCSSCase = (str: string) => {
  return str.split("").reduce(humpToDash, "");
};

const assertStyleMatch = (
  styleSpec: { [key: string]: string | number | undefined },
  component: HTMLElement,
  opts?: jest.Options,
) => {
  Object.entries(styleSpec).forEach(([attr, value]) => {
    expect(component).toHaveStyleRule(toCSSCase(attr), value, opts);
  });
};

const makeArrayKeys = (n: number) => [...Array(n).keys()];

const marginProps = [
  ["m", "margin"],
  ["ml", "marginLeft"],
  ["mr", "marginRight"],
  ["mt", "marginTop"],
  ["mb", "marginBottom"],
  ["mx", "marginLeft"],
  ["mx", "marginRight"],
  ["my", "marginTop"],
  ["my", "marginBottom"],
] as const;

type MarginProps = {
  [K in (typeof marginProps)[number][0]]?: string | number;
};

const paddingProps = [
  ["p", "padding"],
  ["pl", "paddingLeft"],
  ["pr", "paddingRight"],
  ["pt", "paddingTop"],
  ["pb", "paddingBottom"],
  ["px", "paddingLeft"],
  ["px", "paddingRight"],
  ["py", "paddingTop"],
  ["py", "paddingBottom"],
] as const;

type PaddingProps = {
  [K in (typeof paddingProps)[number][0]]?: string | number;
};

const colorProps = [
  ["color", "color", "#CCCCCC"],
  ["bg", "background-color", "#FFFFFF"],
  ["opacity", "opacity", "0.5"],
] as const;

const widthProps = ["width", "width", "200px"] as const;

const heightProps = ["height", "height", "200px"] as const;

const layoutProps = [
  widthProps,
  heightProps,
  ["minWidth", "min-width", "120px"],
  ["maxWidth", "max-width", "120px"],
  ["minHeight", "min-height", "120px"],
  ["maxHeight", "max-height", "120px"],
  ["size", "width", "120px"],
  ["size", "height", "120px"],
  ["display", "display", "inline-block"],
  ["verticalAlign", "vertical-align", "baseline"],
  ["overflow", "overflow", "hidden"],
  ["overflowX", "overflow-x", "hidden"],
  ["overflowY", "overflow-y", "hidden"],
] as const;

const flexBoxProps = [
  ["alignItems", "alignItems", "center"],
  ["alignContent", "alignContent", "center"],
  ["justifyItems", "justifyItems", "center"],
  ["justifyContent", "justifyContent", "center"],
  ["flexWrap", "flexWrap", "wrap"],
  ["flexDirection", "flexDirection", "row-reverse"],
  ["flex", "flex", "1"],
  ["flexGrow", "flexGrow", "1"],
  ["flexShrink", "flexShrink", "1"],
  ["flexBasis", "flexBasis", "100px"],
  ["justifySelf", "justifySelf", "center"],
  ["alignSelf", "alignSelf", "center"],
  ["order", "order", "1"],
] as const;

const gridProps = [
  ["gridColumn", "gridColumn", "1 / span 2"],
  ["gridRow", "gridRow", "1 / span 2"],
  ["gridArea", "gridArea", "myArea"],
  ["gridAutoFlow", "gridAutoFlow", "column"],
  ["gridAutoRows", "gridAutoRows", "150px"],
  ["gridAutoColumns", "gridAutoColumns", "50px"],
  ["gridTemplateRows", "gridTemplateRows", "100px 300px"],
  ["gridTemplateColumns", "gridTemplateColumns", "auto auto auto auto"],
  ["gridTemplateAreas", "gridTemplateAreas", "myArea myArea . . ."],
] as const;

const backgroundProps = [
  [
    "background",
    "background",
    "lightblue url('foo.jpg') no-repeat fixed center",
  ],
  ["backgroundImage", "background-image", "url(foo.jpg)"],
  ["backgroundSize", "background-size", "center"],
  ["backgroundRepeat", "background-repeat", "no-repeat"],
] as const;

const positionProps = [
  ["top", "0px"],
  ["bottom", "0px"],
  ["right", "0px"],
  ["left", "0px"],
  ["position", "fixed"],
  ["position", "sticky"],
  ["position", "absolute"],
  ["position", "static"],
  ["position", "relative"],
] as const;

export const getDefaultValue = (value?: string | number) => {
  const spaceArrayLength = space.length - 1;
  if (value === undefined) return value;
  const parsedValue = +value;
  if (typeof value === "string" && parsedValue > spaceArrayLength) {
    return `${value}px`;
  }
  if (parsedValue <= spaceArrayLength) {
    return space[parsedValue];
  }
  if (parsedValue > spaceArrayLength) {
    return `${parsedValue * 8}px`;
  }
  return value;
};

const testStyledSystemGrid = (
  component: (gridProperties?: GridProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(gridProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component(props));

        assertStyleMatch({ [propName]: value }, elementQuery());
      });
    },
  );
};

const testStyledSystemBackground = (
  component: (backgroundProperties?: BackgroundProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(backgroundProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));

        assertStyleMatch({ [styledSystemProp]: value }, elementQuery());
      });
    },
  );
};

const testStyledSystemPosition = (
  component: (positionProperties?: PositionProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(positionProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, value) => {
      it(`then ${styledSystemProp} should have been set correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));

        assertStyleMatch({ [styledSystemProp]: value }, elementQuery());
      });
    },
  );
};

// this util will catch that a console output occurred without polluting the output when running the unit tests
const expectConsoleOutput = (
  message: string,
  type: "warn" | "error" = "error",
) => {
  if (!message) {
    throw new Error(`no ${type} message provided`);
  }

  expect.assertions(1);

  const consoleType = global.console[type];
  let consoleArgs: string[];

  jest.spyOn(global.console, type).mockImplementation((...args) => {
    if (!args.length) return;

    const msg = args.join(" ");
    const params = args.slice(1, args.length);

    if (sprintf(msg, ...params).includes(message)) {
      consoleArgs = args;
      return;
    }

    consoleType(...args);
  });

  return () => {
    if (consoleArgs) {
      // eslint-disable-next-line no-console
      expect(console[type]).toHaveBeenCalledWith(...consoleArgs);
    }

    global.console[type] = consoleType;
  };
};

const testStyledSystemMargin = (
  component: (spacingProps?: MarginProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: MarginProps,
  assertOpts?: jest.Options,
) => {
  describe("default props", () => {
    let StyleElement: HTMLElement;

    beforeAll(() => {
      render(component({ ...defaults }));
      StyleElement = elementQuery();
    });

    it("should set the correct margins", () => {
      let margin;
      let marginLeft;
      let marginRight;
      let marginTop;
      let marginBottom;

      if (defaults) {
        margin = getDefaultValue(defaults.m);
        marginLeft = getDefaultValue(defaults.ml || defaults.mx);
        marginRight = getDefaultValue(defaults.mr || defaults.mx);
        marginTop = getDefaultValue(defaults.mt || defaults.my);
        marginBottom = getDefaultValue(defaults.mb || defaults.my);

        assertStyleMatch(
          {
            margin,
            marginLeft,
            marginRight,
            marginTop,
            marginBottom,
          },
          StyleElement,
          assertOpts,
        );
      } else {
        expect(StyleElement).not.toHaveStyleRule("marginLeft");
        expect(StyleElement).not.toHaveStyleRule("marginRight");
        expect(StyleElement).not.toHaveStyleRule("marginTop");
        expect(StyleElement).not.toHaveStyleRule("marginBottom");
        expect(StyleElement).not.toHaveStyleRule("margin");
      }
    });
  });

  describe.each(marginProps)(
    'when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: 2 };
        render(component({ ...props }));

        assertStyleMatch(
          { [propName]: "var(--spacing200)" },
          elementQuery(),
          assertOpts,
        );
      });
    },
  );
};

const testStyledSystemPadding = (
  component: (spacingProps?: PaddingProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: PaddingProps,
  assertOpts?: jest.Options,
) => {
  describe("default props", () => {
    let StyleElement: HTMLElement;

    beforeAll(() => {
      render(component({ ...defaults }));
      StyleElement = elementQuery();
    });

    it("should set the correct paddings", () => {
      let padding;
      let paddingLeft;
      let paddingRight;
      let paddingTop;
      let paddingBottom;

      if (defaults) {
        padding = getDefaultValue(defaults.p);
        paddingLeft = getDefaultValue(defaults.pl || defaults.px);
        paddingRight = getDefaultValue(defaults.pr || defaults.px);
        paddingTop = getDefaultValue(defaults.pt || defaults.py);
        paddingBottom = getDefaultValue(defaults.pb || defaults.py);

        assertStyleMatch(
          {
            padding,
            paddingLeft,
            paddingRight,
            paddingTop,
            paddingBottom,
          },
          StyleElement,
          assertOpts,
        );
      } else {
        expect(StyleElement).not.toHaveStyleRule("paddingLeft");
        expect(StyleElement).not.toHaveStyleRule("paddingRight");
        expect(StyleElement).not.toHaveStyleRule("paddingTop");
        expect(StyleElement).not.toHaveStyleRule("paddingBottom");
        expect(StyleElement).not.toHaveStyleRule("padding");
      }
    });
  });

  describe.each(paddingProps)(
    'when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: 2 };
        render(component({ ...props }));

        assertStyleMatch(
          { [propName]: "var(--spacing200)" },
          elementQuery(),
          assertOpts,
        );
      });
    },
  );
};

const testStyledSystemSpacing = (
  component: (spacingProps?: MarginProps | PaddingProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: MarginProps | PaddingProps,
  assertOpts?: jest.Options,
) => {
  testStyledSystemMargin(
    component,
    elementQuery,
    defaults as MarginProps,
    assertOpts,
  );
  testStyledSystemPadding(
    component,
    elementQuery,
    defaults as PaddingProps,
    assertOpts,
  );
};

const testStyledSystemColor = (
  // https://stackoverflow.com/questions/53711454/styled-system-props-typing-with-typescript
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  component: (colorProperties?: any) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(colorProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));
        const StyleElement = elementQuery();
        // Some props need to have camelcase so used toHaveStyleRule rather than assertStyleMatch
        expect(StyleElement).toHaveStyleRule(propName, value);
      });
    },
  );
};

const testStyledSystemWidth = (
  component: (widthProperties?: { width: string }) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe("when a width prop is specified using styled system props", () => {
    it("should set the width styling correctly", () => {
      const [styledSystemProp, propName, value] = widthProps;
      const props = { [styledSystemProp]: value };
      render(component({ ...props }));
      const StyleElement = elementQuery();
      expect(StyleElement).toHaveStyleRule(propName, value);
    });
  });
};

const testStyledSystemHeight = (
  component: (heightProperties?: { height: string }) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe("when a height prop is specified using styled system props", () => {
    it("should set the height styling correctly", () => {
      const [styledSystemProp, propName, value] = heightProps;
      const props = { [styledSystemProp]: value };
      render(component({ ...props }));
      const StyleElement = elementQuery();
      expect(StyleElement).toHaveStyleRule(propName, value);
    });
  });
};

const testStyledSystemLayout = (
  component: (layoutProperties?: LayoutProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(layoutProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));
        const StyleElement = elementQuery();
        // Some props need to have camelcase so used toHaveStyleRule rather than assertStyleMatch
        expect(StyleElement).toHaveStyleRule(propName, value);
      });
    },
  );
};

const testStyledSystemFlexBox = (
  component: (flexboxProperties?: FlexboxProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(flexBoxProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component(props));

        assertStyleMatch({ [propName]: value }, elementQuery());
      });
    },
  );
};

export {
  assertStyleMatch,
  toCSSCase,
  makeArrayKeys,
  mockMatchMedia,
  testStyledSystemColor,
  testStyledSystemLayout,
  testStyledSystemFlexBox,
  testStyledSystemGrid,
  testStyledSystemBackground,
  testStyledSystemPosition,
  expectConsoleOutput,
  testStyledSystemSpacing,
  testStyledSystemMargin,
  testStyledSystemPadding,
  testStyledSystemWidth,
  testStyledSystemHeight,
};
