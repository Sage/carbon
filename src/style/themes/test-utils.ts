const isObject = (obj: Record<string, unknown>) => {
  return typeof obj === "object" && obj !== null;
};

// eslint-disable-next-line import/prefer-default-export
export const assertIsSubset = (
  obj: Record<string, unknown>,
  comparison: Record<string, unknown>,
) => {
  if (!isObject(obj)) {
    // no further nesting
    return;
  }

  const objKeys = Object.keys(obj);
  const comparisonKeys = Object.keys(comparison);

  objKeys.forEach((key) => {
    // assert that keys are present
    expect(comparisonKeys).toContain(key);

    // repeat for nested objects
    assertIsSubset(
      obj[key] as Record<string, unknown>,
      comparison[key] as Record<string, unknown>,
    );
  });
};
