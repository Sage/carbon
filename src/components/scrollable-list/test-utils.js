import React from 'react';
import { makeArrayKeys } from '../../__spec_helper__/test-utils';
import ScrollableListItem from './scrollable-list-item.component';

const listItemReducer = ({ nonSelectables = [], customSelectables = [] }) => {
  // generate jsx for selectable list items, based on indexes in config
  return (acc, item, index) => {
    if (nonSelectables.includes(item)) return [...acc, <div />];
    if (customSelectables.includes(item)) return [...acc, <div isSelectable />];

    return [...acc, <ScrollableListItem id={ index } />];
  };
};

const renderListItems = (opts) => {
  return makeArrayKeys(opts.num).reduce(listItemReducer(opts), []);
};

export default renderListItems;
