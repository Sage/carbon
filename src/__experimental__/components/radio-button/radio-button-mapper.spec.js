import React, { useState } from 'react';
import { act } from 'react-dom/test-utils';
import { ThemeProvider } from 'styled-components';
import TestRenderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import PropTypes from 'prop-types';

import mintTheme from '../../../style/themes/mint';
import RadioButtonMapper from './radio-button-mapper.component';
import { RadioButton } from '.';
import Button from '../../../components/button';

const buttonValues = ['test-1', 'test-2'];
const name = 'test-group';

function render(renderer = TestRenderer.create, props, theme = mintTheme) {
  const children = buttonValues.map((value, index) => (
    <RadioButton
      id={ `rId-${index}` }
      key={ `radio-key-${value}` }
      onChange={ jest.fn() }
      value={ value }
    />
  ));

  return renderer(
    <ThemeProvider theme={ theme }>
      <RadioButtonMapper
        name={ name }
        legend='Test RadioButtonGroup Legend'
        onBlur={ jest.fn() }
        onChange={ jest.fn() }
        { ...props }
      >
        {children}
      </RadioButtonMapper>
    </ThemeProvider>

  );
}

function getRadioButtons(wrapper) {
  return wrapper.find(RadioButton);
}

function getButtons(wrapper) {
  return wrapper.find(Button);
}

function getInputWrapper(button) {
  return button.find('input');
}

describe('RadioButtonMapper', () => {
  it('renders as expected', () => {
    expect(render()).toMatchSnapshot();
  });

  describe('child RadioButton prop / key mapping', () => {
    const wrapper = render(mount);
    const buttons = getRadioButtons(wrapper);
    const buttonArray = buttons.getElements();

    describe.each(buttonArray)('buttons[%#]', (button) => {
      const index = buttonArray.indexOf(button);

      describe('key / value (both derived from value prop)', () => {
        const expectedValue = buttonValues[index];
        const expectedKey = `.$radio-key-${expectedValue}`;

        it(`sets the value to ${expectedValue}`, () => {
          expect(button.props.value).toEqual(expectedValue);
        });

        it(`sets the key to ${expectedKey}`, () => {
          expect(button.key).toEqual(expectedKey);
        });
      });

      describe('name', () => {
        it('is set using the RadioButtonMapper name prop', () => {
          const buttonWrapper = buttons.at(buttonArray.indexOf(button));
          const input = getInputWrapper(buttonWrapper).instance();

          expect(input.name).toEqual(name);
        });
      });
    });

    describe('selected button state', () => {
      describe('initial', () => {
        it('sets checked to false for both buttons', () => {
          buttonArray.forEach((button) => {
            expect(button.props.checked).toBe(false);
          });
        });
      });

      describe('defaultChecked', () => {
        it('sets a child radio button to checked when the prop is set programatically', () => {
          const radioGroup = shallow(
            <RadioButtonMapper
              name={ name }
              legend='Test RadioButtonGroup Legend'
            >
              <RadioButton defaultChecked value='foo' />
            </RadioButtonMapper>
          );

          const button = radioGroup.find(RadioButton);
          expect(button.props().checked).toBe(true);
        });
      });

      describe.each(buttonArray)('when buttons[%#] has changed', (button) => {
        const index = buttonArray.indexOf(button);
        const otherIndex = index ? 0 : 1;
        let buttonWrapper = buttons.at(index);
        let otherButtonWrapper = buttons.at(otherIndex);
        const inputWrapper = getInputWrapper(buttonWrapper);
        const target = inputWrapper.instance();

        inputWrapper.simulate('change', { target });
        wrapper.update();

        buttonWrapper = getRadioButtons(wrapper).at(index);
        otherButtonWrapper = getRadioButtons(wrapper).at(otherIndex);

        it('sets checked === true when it is changed', () => {
          expect(buttonWrapper.props().checked).toBe(true);
          expect(otherButtonWrapper.props().checked).toBe(false);
        });

        it('sets checked === false when the other button is selected', () => {
          const otherInputWrapper = getInputWrapper(otherButtonWrapper);
          const otherTarget = otherInputWrapper.instance();

          otherInputWrapper.simulate('change', { target: otherTarget });
          wrapper.update();

          buttonWrapper = getRadioButtons(wrapper).at(index);
          otherButtonWrapper = getRadioButtons(wrapper).at(otherIndex);

          expect(buttonWrapper.props().checked).toBe(false);
          expect(otherButtonWrapper.props().checked).toBe(true);
        });
      });
    });
  });

  describe('defaultChecked', () => {
    it('sets a child radio button to checked when the prop is set programatically', () => {
      const radioGroup = mount(
        <RadioButtonMapper
          name={ name }
          legend='Test RadioButtonGroup Legend'
        >
          <RadioButton
            defaultChecked
            name='foo'
            value='foo'
          />
          <RadioButton
            name='bar'
            value='bar'
          />
        </RadioButtonMapper>
      );

      const button = getRadioButtons(radioGroup).at(0);
      expect(button.prop('checked')).toBe(true);
    });
  });

  const renderUncontrolled = (groupProps, radioProps) => mount(
    <RadioButtonMapper
      legend='Test RadioButtonGroup Legend'
      name='radio-button-group'
      onChange={ jest.fn() }
      { ...groupProps }
    >
      <RadioButton
        name='one'
        value='one'
      />
      <RadioButton
        name='two'
        value='two'
        { ...radioProps }
      />
      <RadioButton
        name='three'
        value='three'
      />
    </RadioButtonMapper>
  );

  const Controller = (props) => {
    const [value, setValue] = useState(null);
    return (
      <>
        <RadioButtonMapper
          name={ name }
          legend='Test RadioButtonGroup Legend'
          onBlur={ (e) => {
            setValue(e.target.value);
          } }
          onChange={ (e) => {
            setValue(e.target.value);
          } }
          value={ value }
          { ...props.groupProps }
        >
          <RadioButton
            name='one'
            value='one'
          />
          <RadioButton
            name='two'
            value='two'
            { ...props.radioProps }
          />
          <RadioButton
            name='three'
            value='three'
          />
        </RadioButtonMapper>
        <Button onClick={ () => {
          setValue('one');
        } }
        >Set One
        </Button>
        <Button onClick={ () => {
          setValue('two');
        } }
        >Set Two
        </Button>
      </>
    );
  };
  Controller.propTypes = {
    groupProps: PropTypes.any,
    radioProps: PropTypes.any
  };
  const renderControlled = (groupProps = {}, radioProps = {}) => {
    return mount(<Controller { ...{ groupProps, radioProps } } />);
  };


  describe.each([
    ['controlled', renderControlled],
    ['uncontrolled', renderUncontrolled]
  ])('%s', (type, renderer) => {
    it('none of the radio buttons are checked by default', () => {
      const wrapper = renderer();
      const radio = getRadioButtons(wrapper);
      expect(radio.at(0).prop('checked')).toBe(false);
      expect(radio.at(1).prop('checked')).toBe(false);
      expect(radio.at(2).prop('checked')).toBe(false);
    });

    it('onBlur handler is called when a radio button is blurred', () => {
      const onBlur = jest.fn();
      const wrapper = renderer({ onBlur });

      const radio = getRadioButtons(wrapper);

      act(() => {
        radio.at(0).props().onBlur({ target: radio.at(0).getDOMNode() });
      });

      expect(onBlur).toHaveBeenCalled();
    });

    it('onChange handler is called when a radio button is clicked', () => {
      const onChange = jest.fn();
      const wrapper = renderer({ onChange });

      const radio = getRadioButtons(wrapper);

      act(() => {
        radio.at(0).props().onChange({ target: radio.at(0).getDOMNode() });
      });

      expect(onChange).toHaveBeenCalled();
    });
  });

  describe('controlled', () => {
    it('changing the value checks the appropraite radio button', () => {
      const wrapper = renderControlled();
      const buttons = getButtons(wrapper);

      buttons.at(0).simulate('click');

      const radio = getRadioButtons(wrapper);
      expect(radio.at(0).prop('checked')).toBe(true);
      expect(radio.at(1).prop('checked')).toBe(false);
      expect(radio.at(2).prop('checked')).toBe(false);
    });
  });

  describe('uncontrolled', () => {
    it('clicking a value checks the appropraite radio button', () => {
      const wrapper = renderUncontrolled();
      let radio = getRadioButtons(wrapper);

      radio.at(0).find('input').simulate('change');

      radio = getRadioButtons(wrapper);
      expect(radio.at(0).prop('checked')).toBe(true);
      expect(radio.at(1).prop('checked')).toBe(false);
      expect(radio.at(2).prop('checked')).toBe(false);
    });
  });
});
