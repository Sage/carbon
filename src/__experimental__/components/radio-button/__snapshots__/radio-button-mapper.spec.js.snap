// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtonMapper renders as expected 1`] = `
Array [
  .c3 + .c2 {
  margin-top: 16px;
}

.c3.c3.c3 {
  margin-bottom: 16px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 {
  cursor: pointer;
  opacity: 0;
  margin: 0;
  position: relative;
  z-index: 999;
}

.c7 {
  display: inline-block;
  position: relative;
}

.c1 .c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 .c12 {
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding-top: 0;
  width: auto;
}

.c1 .c12 .c13,
.c1 .c12 .c14 {
  color: rgba(0,0,0,0.65);
  vertical-align: middle;
}

.c1 .c12 .c13:hover,
.c1 .c12 .c14:hover,
.c1 .c12 .c13:focus,
.c1 .c12 .c14:focus {
  color: rgba(0,0,0,0.9);
}

.c1 .c11 {
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
}

.c0 {
  margin-top: 8px;
  margin-top: 8px;
}

.c0 .c6 {
  padding-top: 1px;
}

.c0 .c10 {
  height: 16px;
}

.c0 svg {
  background-color: #FFFFFF;
  border: 1px solid #668592;
}

.c0 .c8,
.c0 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c0 .c6,
.c0 .c8,
.c0 .c10,
.c0 svg {
  box-sizing: border-box;
  min-width: 16px;
  width: 16px;
}

.c0 .c8:not([disabled]):focus + .c10,
.c0 .c8:not([disabled]):hover + .c10 {
  box-shadow: 0 0 0 3px #FFB500;
}

.c0 .c12 {
  width: auto;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
}

.c0 .c11 {
  margin-left: 16px;
  margin-top: 0;
  padding-left: 8px;
}

.c0 .c14 {
  position: relative;
  display: inline-block;
}

.c0 .c8:checked ~ .c10 svg path {
  fill: rgba(0,0,0,0.90);
}

.c0 .c10 {
  padding: 0;
}

.c0 .c10,
.c0 svg {
  border-radius: 50%;
}

.c0 .c6,
.c0 .c8,
.c0 .c10,
.c0 svg {
  height: 16px;
  width: 16px;
}

.c0 svg {
  padding: 1px;
}

.c0 circle {
  r: 5;
}

.c0 .c12 {
  -webkit-flex: 1 1 calc(100% - 44px);
  -ms-flex: 1 1 calc(100% - 44px);
  flex: 1 1 calc(100% - 44px);
}

.c0 .c8:checked + .c10 circle {
  fill: rgba(0,0,0,0.9);
}

<div
    checked={false}
    className="c0"
    data-component="radio-button"
    name="test-group"
  >
    <div
      checked={false}
      className="c1"
      name="test-group"
    >
      <div
        className="c2 c3"
      >
        <div
          className="c4 c5"
        >
          <div
            className="c6 c7"
          >
            <input
              aria-checked={false}
              aria-describedby="rId-0-help"
              aria-labelledby="rId-0-label"
              checked={false}
              className="c8 c9"
              id="rId-0"
              name="test-group"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              role="radio"
              type="radio"
              value="test-1"
            />
            <div
              className="c10 "
            >
              <svg
                focusable="false"
                viewBox="0 0 15 15"
              >
                <g
                  fill="none"
                  fillRule="evenodd"
                  stroke="none"
                  strokeWidth="1"
                >
                  <circle
                    className="radio-button-check"
                    cx="7.5"
                    cy="7.5"
                    fill="#FFFFFF"
                    r="5"
                  />
                </g>
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  .c3 + .c2 {
  margin-top: 16px;
}

.c3.c3.c3 {
  margin-bottom: 16px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 {
  cursor: pointer;
  opacity: 0;
  margin: 0;
  position: relative;
  z-index: 999;
}

.c7 {
  display: inline-block;
  position: relative;
}

.c1 .c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 .c12 {
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding-top: 0;
  width: auto;
}

.c1 .c12 .c13,
.c1 .c12 .c14 {
  color: rgba(0,0,0,0.65);
  vertical-align: middle;
}

.c1 .c12 .c13:hover,
.c1 .c12 .c14:hover,
.c1 .c12 .c13:focus,
.c1 .c12 .c14:focus {
  color: rgba(0,0,0,0.9);
}

.c1 .c11 {
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
}

.c0 {
  margin-top: 8px;
  margin-top: 8px;
}

.c0 .c6 {
  padding-top: 1px;
}

.c0 .c10 {
  height: 16px;
}

.c0 svg {
  background-color: #FFFFFF;
  border: 1px solid #668592;
}

.c0 .c8,
.c0 svg {
  height: 16px;
  position: absolute;
  padding: 1px;
}

.c0 .c6,
.c0 .c8,
.c0 .c10,
.c0 svg {
  box-sizing: border-box;
  min-width: 16px;
  width: 16px;
}

.c0 .c8:not([disabled]):focus + .c10,
.c0 .c8:not([disabled]):hover + .c10 {
  box-shadow: 0 0 0 3px #FFB500;
}

.c0 .c12 {
  width: auto;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
}

.c0 .c11 {
  margin-left: 16px;
  margin-top: 0;
  padding-left: 8px;
}

.c0 .c14 {
  position: relative;
  display: inline-block;
}

.c0 .c8:checked ~ .c10 svg path {
  fill: rgba(0,0,0,0.90);
}

.c0 .c10 {
  padding: 0;
}

.c0 .c10,
.c0 svg {
  border-radius: 50%;
}

.c0 .c6,
.c0 .c8,
.c0 .c10,
.c0 svg {
  height: 16px;
  width: 16px;
}

.c0 svg {
  padding: 1px;
}

.c0 circle {
  r: 5;
}

.c0 .c12 {
  -webkit-flex: 1 1 calc(100% - 44px);
  -ms-flex: 1 1 calc(100% - 44px);
  flex: 1 1 calc(100% - 44px);
}

.c0 .c8:checked + .c10 circle {
  fill: rgba(0,0,0,0.9);
}

<div
    checked={false}
    className="c0"
    data-component="radio-button"
    name="test-group"
  >
    <div
      checked={false}
      className="c1"
      name="test-group"
    >
      <div
        className="c2 c3"
      >
        <div
          className="c4 c5"
        >
          <div
            className="c6 c7"
          >
            <input
              aria-checked={false}
              aria-describedby="rId-1-help"
              aria-labelledby="rId-1-label"
              checked={false}
              className="c8 c9"
              id="rId-1"
              name="test-group"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              role="radio"
              type="radio"
              value="test-2"
            />
            <div
              className="c10 "
            >
              <svg
                focusable="false"
                viewBox="0 0 15 15"
              >
                <g
                  fill="none"
                  fillRule="evenodd"
                  stroke="none"
                  strokeWidth="1"
                >
                  <circle
                    className="radio-button-check"
                    cx="7.5"
                    cy="7.5"
                    fill="#FFFFFF"
                    r="5"
                  />
                </g>
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
